tests/models/test_causal_self_attention.py

```diff
@@ -60,14 +121,102 @@ def test_forward_pass_success(n_head_q, n_head_kv, n_embd, successful):
+    "n_head_q, n_head_kv, n_embd, attention_impl_1, attention_impl_2, verbose",
+    [
+        # manual vs. pytorch_flash
+        (4, 4, 4, "manual", "pytorch_flash", False),  # MHA
+        (4, 4, 32, "manual", "pytorch_flash", False),
+        (4, 4, 768, "manual", "pytorch_flash", False),
+        (8, 8, 2048, "manual", "pytorch_flash", False),
+        (8, 2, 2048, "manual", "pytorch_flash", False),  # GQA
+        # manual vs. dao_flash
+        (4, 4, 4, "manual", "dao_flash", False),  # MQA
+        (4, 4, 32, "manual", "dao_flash", False),
+        (4, 4, 768, "manual", "dao_flash", False),
+        (8, 8, 2048, "manual", "dao_flash", False),
+        (8, 2, 2048, "manual", "dao_flash", False),  # GQA
+        # pytorch_flash vs. dao_flash
+        (4, 4, 4, "pytorch_flash", "dao_flash", False),
+        (4, 4, 32, "pytorch_flash", "dao_flash", False),
+        (4, 4, 768, "pytorch_flash", "dao_flash", False),
+        (8, 8, 2048, "pytorch_flash", "dao_flash", False),
+        (8, 2, 2048, "pytorch_flash", "dao_flash", False),  # GQA
+    ],
+)
+def test_attention_implementation_approximate_equality(
+    n_head_q, n_head_kv, n_embd, attention_impl_1, attention_impl_2, verbose
+):
+    # flash attention is non-deterministic,
+    # see https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
+    # and https://pytorch.org/docs/stable/notes/randomness.html#avoiding-nondeterministic-algorithms
+    # as well as https://github.com/pytorch/pytorch/issues/119188#issuecomment-2043157422
+
+    batch_size = 2
+    block_size = 10
+    embedding_shape = (batch_size, block_size - 1, n_embd)
+    embedded_input_seq = _get_random_input_seq(embedding_shape)
+
+    attention_config = AttentionConfig(qkv_transforms=[])
+    attention_layer_args = {
+        "n_head_q": n_head_q,
+        "n_head_kv": n_head_kv,
+        "n_embd": n_embd,
+        "attention_config": attention_config,
+        "attention_impl": attention_impl_1,
+    }
 
-    out_flash = CausalSelfAttention.execute_flash_attention(query_orig, key_orig, value_orig, dropout=0.0)
+    attention_layer = {}
+    attention_layer[attention_impl_1] = _get_random_attention_layer(**attention_layer_args)
+    attention_layer[attention_impl_2] = deepcopy(attention_layer[attention_impl_1])
+    attention_layer[attention_impl_2].attention_impl = attention_impl_2
 
-    assert out_flash.shape == (1, 12, 8, 16)
+    output_tensor = {}
+    output_tensor[attention_impl_1] = attention_layer[attention_impl_1](embedded_input_seq)
+    output_tensor[attention_impl_2] = attention_layer[attention_impl_2](embedded_input_seq)
+    if verbose:
+        print(f"{attention_impl_1} vs. {attention_impl_2}: \n{output_tensor}")
+    torch.testing.assert_close(
+        output_tensor[attention_impl_1],
+        output_tensor[attention_impl_2],
+        atol=2.5e-3,  # default for bfloat16: 1e-5
+        rtol=0.016,  # default for bfloat16: 0.016
+    )
```

Comment:
We could add another test in which we test the output of the  manual attention implementation against the precomputed output (pen and paper) given a very short input sequence. In this case, we can be entirely sure that the implementation is correct.